Feature/farm 280 ui update discover

jessicahernandezhv opened the pull request:
[[FARM-280 - UI update: discover](https://amidodevelopment.atlassian.net/secure/RapidBoard.jspa?rapidView=150&projectKey=FARM&view=planning&selectedIssue=FARM-280)]

#### 📲 What

Applied new UI design in Discover screen. Refactor code.

#### 🤔 Why
		
Because we need to implement new UI style.
		
#### 🛠 How

* New UI design with new widgets		
* Externalised widgets
* Local style/theme variables
* Code refactor
* TODO: update typography

#### 👀 See
		
![discover_redesing](https://user-images.githubusercontent.com/34380229/59365292-9ccee300-8d38-11e9-824c-502acb6a829f.gif)
		 
#### ✅ Acceptance criteria

- [ ] Design Review for UI with BA completed. 
- [x] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [ ] Passing all tests
- [x] Rebased/merged with latest changes from development and re-tested
- [x] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [ ] unit tests 80% coverage on testable code (functions, methods, class)
- [x] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
- [ ] ui elements defined under lib/ui. 
- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
- [ ] Colors defined in lib/utils/colors.dart
- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
- [ ] TextStyles defined in lib/utils/styles.dart

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo


#### Reviewers

@LeeMatthewHiggins 
@farmsmart/wamf @farmsmart/amido

Commit:
[FARM-280] Basic structure to redesign list view of discover screen

Commit:
[FARM-340] First changes in UI redesign

Commit:
[FARM-240] Adding new constants

Commit:
[FARM-340] Refactor code for Discover page

maribel4amido commented:
I did a quick exploratory test on the last apk build and found some bugs. 
* The first article is not clickable at all. 
* The clickbox should be the entire area of the article (image and text including). If a click on the space between text and image, nothing happens.

maribel4amido commented:
Another question, which User/story or PR will update the icons for the bottom navigation?

LeeMatthewHiggins commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -7,81 +7,58 @@ import 'package:farmsmart_flutter/utils/styles.dart';
 
+import 'discover_page.dart';
+
 class DiscoveryListItem {
```
naming here: I'm asking myself; what are we rendering? - I think a list of articles... So I would name this ArticleListItem.   The fact its in a discovery section is a design detail we may change (maybe we want the articles on the home screen after user testing).

LeeMatthewHiggins replied:
note: remember to rename all the other things that have discover..

jessicahernandezhv replied:
Cool 👍🏻renaming

LeeMatthewHiggins commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -7,81 +7,58 @@ import 'package:farmsmart_flutter/utils/styles.dart';
-          child: Container(
-              height: 101,
-              child: Column(
+      onTap: () => getRelatedArticles(articleData),
```
I don't think its this lists job to get related articles. I think the article detail widget should handle that. what if we change related articles or removed them? We may get to an article from another entry point, we would have to add the fetch at all the entry points, its better to let the article widget (the detail view) fetch repeated articles, then its self contained and can work without any dependences on other widgets fetching data it needs....

jessicahernandezhv replied:
We are doing the fetch of related articles here to handle de loading status of Article detail screen. When user taps an article, we fetch the related articles of this specific article and present the detail screen in Loading status, once we have the data the status changes to success.

LeeMatthewHiggins replied:
Lets change this to fetch in the article widget, and just make this show the article widget in loading state, the article widget should then trigger the fetch on init. We can then rename this function to showDetail

LeeMatthewHiggins commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -7,81 +7,58 @@ import 'package:farmsmart_flutter/utils/styles.dart';
-        height: listImageHeight,
-        width: listImageWidth,
-        fit: BoxFit.cover
+  _buildListItemImage(ArticleEntity articleData, HomeDiscoverPageStyle discoverStyle) {
```
As with @isaacmartinezimv ´s PR. Let´s introduce view models.
since the widget is very much like a generic table view cell, we could have a very generic model here:
SimpleListItemViewModel
-title
-subtitle
-thumbnailURI

LeeMatthewHiggins commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -7,81 +7,58 @@ import 'package:farmsmart_flutter/utils/styles.dart';
   }
 
-  _buildArticleTitle(ArticleEntity articleData) {
+  _buildArticleInformation(ArticleEntity articleData, HomeDiscoverPageStyle discoverStyle) {
```
good style injection, swap out the model object for a viewModel and were looking good 👍

LeeMatthewHiggins commented on lib/ui/discover/discover_page.dart:
```diff
@@ -1,13 +1,91 @@
+
+  final EdgeInsets titlePadding;
+  final EdgeInsets headerArticlePadding;
+  final EdgeInsets itemListPadding;
```
lets extract out the list items styles to the list item widget and just have an entry here for the list item style

jessicahernandezhv replied:
Now I split it in two styles, one for ArticleList and other one for ArticleListItem

LeeMatthewHiggins commented on lib/ui/discover/discover_page.dart:
```diff
@@ -16,33 +94,103 @@ class HomeDiscoverPage extends StatefulWidget {
+    );
+  }
 
+  Widget _buildArticlesList(List<ArticleEntity> articlesList, Function getRelatedArticles, HomeDiscoverPageStyle discoverStyle) {
```
let push the responsibility of repeated article fetch to the article detail widget not this list widget.

LeeMatthewHiggins commented on lib/ui/discover/discover_page.dart:
```diff
@@ -16,33 +94,103 @@ class HomeDiscoverPage extends StatefulWidget {
+        physics: ScrollPhysics(),
+        itemCount: articlesList.length-1,
+        itemBuilder: (BuildContext context, int index) {
+          return DiscoveryListItem().buildListItem(articlesList[index+1], getRelatedArticles, discoverStyle);
```
don't like the +1´s and -1´s. I think we should just have a list, but use a different item builder function for the 2 styles.... if index == 0 use heroListItemBuilder else use standardListItemBuilder. This would remove the complexity in the widget building in the above code too.

jessicahernandezhv replied:
Awesome idea! This solves the problem more elegantly

LeeMatthewHiggins reviewed (changes requested):
Good start! Well done.
Lets add in the viewModels and just do a small refactor to swap out the list item builder for the 2 item styles. I would name the large style "HeroXxx" and normal style "standardXxx".

LeeMatthewHiggins commented:
> Another question, which User/story or PR will update the icons for the bottom navigation?

Good question.

I feel like since we need to address that bottom UI component to make sure it has the correct navigation stack ETC. it should be done in a bottom bar navigation task outside this one. (this widget should not know about its container and so we should follow that pattern in this PR since its specific to the article list widget)

LeeMatthewHiggins commented:
> I did a quick exploratory test on the last apk build and found some bugs.
> 
> * The first article is not clickable at all.
> * The clickbox should be the entire area of the article (image and text including). If a click on the space between text and image, nothing happens.

this first article is a special case, I have commented that it shouldn't be, it not being clickable should be fixed with the changes I have suggested on this PR. And you are right @maribel4amido, the whole list item should be clickable @jessicahernandezhv .

maribel4amido commented:
> > Another question, which User/story or PR will update the icons for the bottom navigation?
> 
> Good question.
> 
> I feel like since we need to address that bottom UI component to make sure it has the correct navigation stack ETC. it should be done in a bottom bar navigation task outside this one. (this widget should not know about its container and so we should follow that pattern in this PR since its specific to the article list widget)

Hi Lee. Noted. I've asked @harrylusk  to create a story for the bottom navigation bar. He did say the icons are being finalised as well so hopefully those should be ready when bottom navigation bar can be worked on.

Please note that there is a [JIRA](https://amidodevelopment.atlassian.net/browse/FARM-326) ticket regarding navigation bug. Do we keep the UI changes separate to the navigation bug fixes ?

LeeMatthewHiggins commented:
> > > Another question, which User/story or PR will update the icons for the bottom navigation?
> > 
> > 
> > Good question.
> > I feel like since we need to address that bottom UI component to make sure it has the correct navigation stack ETC. it should be done in a bottom bar navigation task outside this one. (this widget should not know about its container and so we should follow that pattern in this PR since its specific to the article list widget)
> 
> Hi Lee. Noted. I've asked @harrylusk to create a story for the bottom navigation bar. He did say the icons are being finalised as well so hopefully those should be ready when bottom navigation bar can be worked on.
> 
 ATM it´s not really a "bug", its just missing, we have 4 buttons at the bottom of the screen and not a bottom navigation bar. So I think a ticket to just implement the bottom navigation widget would cover the icon updates and the behaviour.

harrylusk commented:
> > > > Another question, which User/story or PR will update the icons for the bottom navigation?
> > > 
> > > 
> > > Good question.
> > > I feel like since we need to address that bottom UI component to make sure it has the correct navigation stack ETC. it should be done in a bottom bar navigation task outside this one. (this widget should not know about its container and so we should follow that pattern in this PR since its specific to the article list widget)
> > 
> > 
> > Hi Lee. Noted. I've asked @harrylusk to create a story for the bottom navigation bar. He did say the icons are being finalised as well so hopefully those should be ready when bottom navigation bar can be worked on.
> 
> ATM it´s not really a "bug", its just missing, we have 4 buttons at the bottom of the screen and not a bottom navigation bar. So I think a ticket to just implement the bottom navigation widget would cover the icon updates and the behaviour.

@LeeMatthewHiggins @maribel4amido I've added a ticket to Jira. Would be good to get this created sooner than later - perhaps this sprint. I'll chat with Jessica & Isaac about this.

Commit:
[PR Comment] Some rename and split styles for widgets

Commit:
[PR Comments] Renaming, implement clickable cards

Commit:
[PR Comments] Renaming variables

Commit:
[PR Comment] Applying temporal view model approach (pending to apply in Detail Screen)

LeeMatthewHiggins commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -7,81 +7,104 @@ import 'package:farmsmart_flutter/utils/styles.dart';
-        height: listImageHeight,
-        width: listImageWidth,
-        fit: BoxFit.cover
+  _buildListItemImage(ArticlesListViewModel articleData, ArticleItemStyle itemStyle) {
```
lets call this variable just "viewModel"

Commit:
First attempt to implement specific view model

Commit:
View model working with functions

Commit:
[PR Comments] View model working on heroItem

Commit:
[FARM-280] Cleaning unused elements

Commit:
[PR Comment] Renamed variables

Commit:
[PR Comments] Renaming review

maribel4amido commented on lib/utils/styles.dart:
```diff
@@ -39,7 +39,11 @@ abstract class Styles {
 
   static TextStyle articleListTitleStyle() {
-    return TextStyle(fontSize: 15, fontWeight: FontWeight.bold, color: Color(black), decorationColor: Color(primaryGreen));
+    return TextStyle(fontSize: 17, fontWeight: FontWeight.w700, color: Color(darkBlue));
```
I see that styles are now created per widget, are these styles (and associated global colors) still in use?

jessicahernandezhv replied:
Global colors are now unused, when we'll migrate all the files into new coding style, this global colors are going to be deleted, or refactor to have only the basic ones

maribel4amido reviewed (dismissed):
Minor points

Commit:
[PR Comments] Split list builders between hero and standard. Solved spacing between cards

LeeMatthewHiggins commented on lib/ui/discover/HeroListItem.dart:
```diff
@@ -0,0 +1,46 @@
+import 'discover_page.dart';
+
+class HeroListItem {
+  Widget builder(ArticlesItemListViewModel viewModel, {ArticleListItemStyle itemStyle = const HeroArticleListItemStyle()}) {
```
this should be _DefaultListItem style and have the default defined in this file.

LeeMatthewHiggins commented on lib/ui/discover/StandardListItem.dart:
```diff
@@ -0,0 +1,142 @@
+  const _DefaultArticleListItemStyle();
+}
+
+class HeroArticleListItemStyle implements ArticleListItemStyle {
```
the hero style in the StandardListItem.dart? doesn't make sense. Would be better to lift this into the Hero widget file and use it as the private _DefaultListItemStyle.

LeeMatthewHiggins commented on lib/ui/discover/discover_page.dart:
```diff
@@ -1,48 +1,136 @@
+
+
+
+// TODO: Maybe we have to externalice this widget in other file
```
this needs its own file.

LeeMatthewHiggins reviewed (changes requested):
One last pass to clean the hero and standard item split and to lift out the divider for reuse

LeeMatthewHiggins committed:
New list view widget with the concept of a header and footer widget.
This will build the list items on demand and remain compact in terms of resources

LeeMatthewHiggins committed:
PR naming consistency
Widget style defaults consistency

LeeMatthewHiggins committed:
Minor naming

LeeMatthewHiggins committed:
Merge commit '88ffe30f51285628bb7972cd24c20518bf6e17ac' into feature/FARM-280-UI-Update-Discover
Sync with Dev

LeeMatthewHiggins committed:
Add in some list view parameters to the headers and footer list

LeeMatthewHiggins committed:
Split out the list item style, as its used by multiple widgets
Split out the list divider widget for reuse
Minor name changes

NB: Attempted to fix the height of the hero image to prevent layout popping when changing from placeholder to real image, but requires a refactor of the network image class.

LeeMatthewHiggins reviewed (dismissed):
This needs a fixed height adding to the hero article list item, but that needs to be done as part of another refactor of the image downloading. For now the hero image height will depend on the source image aspect, which might look odd with some images and will jump when the placeholder and real image is swapped.

LeeMatthewHiggins committed:
Fix for case issue with import.

LeeMatthewHiggins committed:
Merge branch 'develop' into feature/FARM-280-UI-Update-Discover

# Conflicts:
#	lib/ui/home.dart

Also cleaned up imports
